src/proc/validator.rs

```diff
@@ -1005,15 +1140,222 @@ impl Validator {
                 index,
             } => {}
             E::ImageQuery { image, query } => {}
-            E::Unary { op, expr } => {}
-            E::Binary { op, left, right } => {}
+            E::Unary { op, expr } => {
+                use crate::UnaryOperator as Uo;
+                let inner = resolver.resolve(expr)?;
+                match (op, inner.scalar_kind()) {
+                    (_, Some(Sk::Sint))
+                    | (_, Some(Sk::Bool))
+                    | (Uo::Negate, Some(Sk::Float))
+                    | (Uo::Not, Some(Sk::Uint)) => {}
+                    other => {
+                        log::error!("Op {:?} kind {:?}", op, other);
+                        return Err(ExpressionError::InvalidUnaryOperandType(op, expr));
+                    }
+                }
+            }
+            E::Binary { op, left, right } => {
+                use crate::BinaryOperator as Bo;
+                let left_inner = resolver.resolve(left)?;
+                let right_inner = resolver.resolve(right)?;
+                let good = match op {
+                    Bo::Add | Bo::Subtract | Bo::Divide | Bo::Modulo => match *left_inner {
+                        Ti::Scalar { kind, .. } | Ti::Vector { kind, .. } => match kind {
+                            Sk::Uint | Sk::Sint | Sk::Float => left_inner == right_inner,
+                            Sk::Bool => false,
+                        },
+                        _ => false,
+                    },
+                    Bo::Multiply => {
+                        let kind_match = match left_inner.scalar_kind() {
+                            Some(Sk::Uint) | Some(Sk::Sint) | Some(Sk::Float) => true,
+                            Some(Sk::Bool) | None => false,
+                        };
+                        //TODO: should we be more restrictive here? I.e. expect scalar only to the left.
```

Comment:
I would say yes, because it can be crucial for the back-ends processing. However, not really needed for now I guess. It is better to push this validation forward first, and iron out the minor things later.

Reply:
Agreed on both accounts!